rclpy/src/rclpy/_rclpy_signal_handler.c

```diff
@@ -0,0 +1,316 @@
+  PyObject * pygc;
+  if (!PyArg_ParseTuple(args, "O", &pygc)) {
+    return NULL;
+  }
+
+  if (0 != check_signal_safety()) {
+    // exception raised
+    return NULL;
+  }
+
+  rcl_guard_condition_t * gc = (rcl_guard_condition_t *)PyCapsule_GetPointer(
+    pygc, "rcl_guard_condition_t");
+  if (!gc) {
+    return NULL;
+  }
+
+  rcl_guard_condition_t ** guard_conditions;
+  rcutils_atomic_load(&g_guard_conditions, guard_conditions);
+
+  // Figure out how big the list currently is
+  size_t count_gcs = 0;
+  if (NULL != guard_conditions) {
+    while (NULL != guard_conditions[count_gcs]) {
+      if (gc == guard_conditions[count_gcs]) {
+        PyErr_Format(PyExc_ValueError, "Guard condition was already registered");
+        return NULL;
+      }
+      ++count_gcs;
+    }
+  }
+
+  // Current size of guard condition list: count_gcs + 1 (sentinel value)
+  // Allocate space for one more guard condition: count_cs + 1 (new gc) + 1 (sentinel value)
+  rcl_allocator_t allocator = rcl_get_default_allocator();
+  rcl_guard_condition_t ** new_gcs =
+    allocator.allocate(sizeof(rcl_guard_condition_t *) * (count_gcs + 2), allocator.state);
+
+  // populate the new guard condition list, ending with a sentinel of NULL
+  for (size_t i = 0; i < count_gcs; ++i) {
+    new_gcs[i] = guard_conditions[i];
+  }
+  new_gcs[count_gcs] = gc;
+  new_gcs[count_gcs + 1] = NULL;
+
+  // Swap the lists and free the old
+  rcl_guard_condition_t ** old_gcs;
+  rcutils_atomic_exchange(&g_guard_conditions, old_gcs, new_gcs);
+  if (NULL != old_gcs) {
+    allocator.deallocate(old_gcs, allocator.state);
+  }
+
+  // make sure our signal handler is registered
+  register_signal_handler();
+
+  Py_RETURN_NONE;
+}
+
+/// Unregister a guard condition so it is not triggered when SIGINT is received.
+/**
+ * Raises ValueError if the argument is not a guard condition handle
+ * Raises ValueError if the argument was not registered
+ *
+ * \param[in] pygc a guard condition pycapsule
+ * \return None
+ */
+static PyObject *
+rclpy_unregister_sigint_guard_condition(PyObject * Py_UNUSED(self), PyObject * args)
+{
+  // Expect a pycapsule with a guard condition
+  PyObject * pygc;
+  if (!PyArg_ParseTuple(args, "O", &pygc)) {
+    return NULL;
+  }
+
+  rcl_guard_condition_t * gc = (rcl_guard_condition_t *)PyCapsule_GetPointer(
+    pygc, "rcl_guard_condition_t");
+  if (!gc) {
+    return NULL;
+  }
+
+  rcl_guard_condition_t ** guard_conditions;
+  rcutils_atomic_load(&g_guard_conditions, guard_conditions);
+
+  // Figure out how big the list currently is
+  size_t count_gcs = 0;
+  bool found_gc = false;
+  // assumes guard condition was only added to list once
+  size_t found_index = 0;
+
+  if (NULL != guard_conditions) {
+    while (NULL != guard_conditions[count_gcs]) {
+      if (gc == guard_conditions[count_gcs]) {
+        found_gc = true;
+        found_index = count_gcs;
+      }
+      ++count_gcs;
+    }
+  }
+
+  if (count_gcs == 0 || !found_gc) {
```

Comment:
Nitpick: isn't the second condition sufficient on its own?

Reply:
It is, removed first part of condition in 3d25d7f3efe8fe96f72d9c5e41c454a903d7722a